Add builtin method: provision_execute

kbrock opened the pull request:
dependency:
- [x] https://github.com/ManageIQ/manageiq-providers-workflows/pull/85
- [x] https://github.com/ManageIQ/manageiq/pull/23021

status: sharing my current code

Fryguy commented on lib/manageiq/providers/workflows/builtin_methods.rb:
```diff
@@ -21,6 +21,43 @@ def self.email(params = {}, _secrets = {}, _context = {})
+          case request_task.statemachine_task_status
+          when "error"
+            reason = task.message
+            reason = reason[7..] if reason[0..6] == "Error: "
```
```suggestion
            reason = reason[7..] if reason.start_with?("Error: ")
```

or, perhaps just use sub which I think is more readable

```suggestion
            reason = reason.sub(/^Error: ", "")
```

Fryguy commented on lib/manageiq/providers/workflows/builtin_methods.rb:
```diff
@@ -21,6 +21,43 @@ def self.email(params = {}, _secrets = {}, _context = {})
+          when "ok"
+            runner_context["running"] = false
+            runner_context["success"] = false
+            runner_context["output"] = {"Result" => "provisioned"}
```
Should there be an else case?

Fryguy reviewed (dismissed):
Overall LGTM.  @agrare Can you please also review?

Fryguy commented:
@kbrock This PR was titled "Builtin runner accepting context", but then it also includes provision_execute and some variable renames and builtin flow changes - was this supposed to be separate PRs or should the title be renamed to something like "Add provision_execute as a builtin method"?

Fryguy commented:
I just realized this requires https://github.com/ManageIQ/manageiq/pull/23021, so I updated the OP

agrare commented on lib/manageiq/providers/workflows/builtin_methods.rb:
```diff
@@ -21,6 +21,43 @@ def self.email(params = {}, _secrets = {}, _context = {})
+            runner_context["output"] = {"Error" => "States.Timeout", "Cause" => "retry"}
+          when "ok"
+            runner_context["running"] = false
+            runner_context["success"] = false
```
Doesn't `"ok"` mean that success should be true here?

agrare commented on lib/manageiq/providers/workflows/builtin_runner.rb:
```diff
@@ -49,6 +54,9 @@ def success?(runner_context)
         def output(runner_context)
           runner_context["output"]
         end
+
+        # def wait(_timeout: nil, _events: %i[create update delete])
+        # end
```
```suggestion
```

kbrock replied:
?

Fryguy commented on lib/manageiq/providers/workflows/builtin_runner.rb:
```diff
@@ -34,8 +38,9 @@ def wait(timeout: nil, events: %i[create update delete])
+          return if !runner_context["running"]
 
-          BuiltinMethods.send("#{method}_status!", runner_context)
+          BuiltinMethods.send(status_method, runner_context)
```
Where is this `status_method` defined?

Fryguy commented on lib/manageiq/providers/workflows/builtin_methods.rb:
```diff
@@ -21,6 +21,41 @@ def self.email(params = {}, _secrets = {}, _context = {})
+            runner_context["success"] = false
+            runner_context["output"] = {"Error" => "States.TaskFailed", "Cause" => reason}
+          when "retry"
+            runner_context["running"] = false
```
I'm not sure what retry implies, but I think it should still be "running"?

agrare replied:
Oh yeah good point, does "retry" here mean we want the ASL workflow to retry or it is the backend doing the retry?

kbrock replied:
I was thinking that it was the same thing as a failure, the difference being that we will revisit this.

see how this is written in the floe:
https://github.com/ManageIQ/manageiq-content/pull/746

I agree, this suggests that it is a retry, but this version of `retry` may be more like a `wait`.

kbrock replied:
confused by this.
agree with `running=true`, but it is not complete. it is basically a `wait` state so then `success` is not fully determined.

agrare replied:
> agree with running=true, but it is not complete.

Do you mean you don't think we should set success=true?

This could apply to all of these, but in general the container_runners don't set success=true/false until after it has completed so if that's what you mean then yes I agree but then still not sure why we'd say running=false,success=false because that is an error case.

Fryguy replied:
I was thinking running=true and success is not yet determined (not sure if that should be false or just not set at all?)

Fryguy commented on lib/manageiq/providers/workflows/builtin_methods.rb:
```diff
@@ -21,6 +21,41 @@ def self.email(params = {}, _secrets = {}, _context = {})
+
+        private_class_method def self.provision_execute_status!(runner_context)
+          request_task = ::MiqRequestTask.find_by(:id => runner_context["miq_request_task_id"])
+          return if request_task.nil?
```
Regardless, the caller does `method_result.merge(runner_context)`, so this will just cause the caller to blow up with a `undefined method `merge' for nil:NilClass`

agrare replied:
Right I'm thinking we should return `{"running" => false, "success" => "false", "output" => {"Error" => "States.TaskFailed", "Cause" => "Unable to find miq_request_task"}}` or something like that

Fryguy replied:
Oh no wait, I thought this was the execute method as opposed to the status method.  Ignore my comment.

kbrock replied:
this nil is copied from email status.
I'll fix, but maybe we want to consider fixing that other one as well

agrare replied:
>  I'll fix, but maybe we want to consider fixing that other one as well

Yes I think I commented over in the PR that added that but didn't get to fix it, wanted to get it right here

agrare replied:
@kbrock https://github.com/ManageIQ/manageiq-providers-workflows/pull/87/files#diff-f49bf1f6a21ddda0e802d7717ef72ecaef1b11b86ee27fb5db029c7f770e3b0cR19

Fryguy commented on lib/manageiq/providers/workflows/builtin_runner.rb:
```diff
@@ -27,13 +27,10 @@ def cleanup(runner_context)
         def status!(runner_context)
           method = runner_context["method"]
           raise ArgumentError if method.nil?
+          return if !runner_context["running"]
```
I'm not sure I understand this bit - Is this to prevent status being called twice?

kbrock replied:
yes. status! is called 2-3 times once a task is complete

also, if a task fails in the original run, then we don't want it calling status

agrare replied:
> also, if a task fails in the original run, then we don't want it calling status

In the container_runner classes we return if an error has already been set in the runner context, would that work for this case?

kbrock replied:
The check on Error was the inspiration.
The thinking was that if we set the `output` / `running=false`, then we were done with this task and no reason to run status to check further

agrare commented on lib/manageiq/providers/workflows/builtin_methods.rb:
```diff
@@ -21,6 +21,41 @@ def self.email(params = {}, _secrets = {}, _context = {})
+
+          ::MiqRequestTask.find(context.execution["_object_id"])&.execute
+
+          {"miq_request_task_id" => request_task_id}
```
```
#<NameError: undefined local variable or method `request_task_id' for ManageIQ::Providers::Workflows::BuiltinMethods:Class

          {"miq_request_task_id" => request_task_id}
```

Fryguy replied:
Think this is available in line 28 now? (old comment?)

agrare replied:
This comment was from before the most recent push which fixed it

kbrock commented:
update
- fixed miq_request_task_id
- fixed delay
- `s/method/method_name` to avoid using common method name

kbrock commented:
update:
- split into multiple commits
- extract status check for miq_request_task
- move error and status from methods to runner

update:
- fixed `case when nil` case

agrare commented on lib/manageiq/providers/workflows/builtin_methods.rb:
```diff
@@ -14,9 +14,27 @@ def self.email(params = {}, _secrets = {}, _context = {})
+            return BuiltinRunner.error!(runner_context, :cause => "Calling provision_execute on non-provisioning request: [#{runner_context["_object_type"]}]")
+          end
+
+          request_task_id = context.execution["_object_id"]&.to_i
+          ::MiqRequestTask.find(request_task_id)&.execute
```
@kbrock it looks like we go through some effort to ensure `context.execution["_object_type"]` is `"ServiceTemplateProvisionTask"` but then also go through a lot of effort to handle a missing `_object_id`.

Should we also be raising an error if `context.execution["_object_type"]` is missing and if `::MiqRequestTask.find(request_task_id)` can't be found?

agrare commented on lib/manageiq/providers/workflows/builtin_methods.rb:
```diff
@@ -25,17 +43,29 @@ def self.email(params = {}, _secrets = {}, _context = {})
+            runner_context["running"] = true
+            runner_context
+          when "ok"
+            BuiltinRunner.success!(runner_context, :output => {"Result" => "provisioned"})
```
We should probably return something that subsequent states could make use of, maybe the resulting `MiqRequestTask#destination` which is a `Service`

Example use case, run some automation against the vm(s) that get created.

agrare reviewed (approved):
Just a couple of nits that we can do in a follow-up
